server/src/main/java/org/opensearch/rest/RestController.java

```diff
@@ -588,4 +600,47 @@ private static CircuitBreaker inFlightRequestsBreaker(CircuitBreakerService circ
         // We always obtain a fresh breaker to reflect changes to the breaker configuration.
         return circuitBreakerService.getBreaker(CircuitBreaker.IN_FLIGHT_REQUESTS);
     }
+
+    /**
+     * Authenticates the subject of the incoming REST request based on the auth header
+     * @param request the request whose subject is to be authenticated
+     * @param channel the channel to send the response on
+     * @throws IOException when an exception is raised writing response to channel
+     */
+    private void authenticate(RestRequest request, RestChannel channel) throws IOException {
+
+        final Optional<String> authHeader = request.getHeaders()
+            .getOrDefault(HttpHeaderToken.HEADER_NAME, Collections.emptyList())
+            .stream()
+            .findFirst();
+
+        Subject subject = null;
+
+        AuthenticationToken headerToken = null;
+
+        // TODO: Handle anonymous Auth - Allowed or Disallowed (set by the user of the system) - 401 or Login-redirect ??
+
+        if (authHeader.isPresent()) {
+            try {
+                // support other type of header tokens
+                headerToken = new HttpHeaderToken(authHeader.get());
+                subject = Identity.getAuthManager().getSubject();
```

Comment:
I've seen this call in Shiro documentation, but I'm not quite following how this resolves to the correct subject here. 

According to Shiro docs:

```
In almost all environments, you can obtain the currently executing Subject by using org.apache.shiro.SecurityUtils:

Subject currentUser = SecurityUtils.getSubject();

The getSubject() call in a standalone application might return a Subject based on user data in an application-specific location, and in a server environment (e.g. web app), it acquires the Subject based on user data associated with current thread or incoming request.
```

https://shiro.apache.org/subject.html

Will this always resolve to the correct Subject?

Reply:
yes.. Shiro has its own resource store and it fetches correct subject based on the store:

This is how shiro puts resource into its ThreadContext store:

https://github.com/apache/shiro/blob/main/core/src/main/java/org/apache/shiro/util/ThreadContext.java#L153

Reply:
According to shiro source, the DefaultSecurityManager uses ThreadContext under the hood: https://github.com/apache/shiro/blob/main/core/src/main/java/org/apache/shiro/SecurityUtils.java#LL54C9-L54C54

Reply:
I'm not sure I understand the ask here